### PR TITLE
Adding storeLargeFileFromLocalContentAsync method to B2StorageClient

### PR DESCRIPTION
**Pull request by malaysf**
This behaves like `storeLargeFileFromLocalContent` but returns a cancellable future which will prevent additional tasks from starting and attempt to cancel in-progress API calls.

Addresses #16

### CODE REVIEW

**Review by certainmagic — COMMENTED** [review]
so far, i've just written some nitpicks.  still letting the overall thing sink in.  (and i skimmed the completion stuff.)

File: core/src/main/java/com/backblaze/b2/client/B2CancellationToken.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Token to pass around whether the caller has cancelled an operation so that
+ * sub tasks can stop their processing.
+ */
+class B2CancellationToken {
+    /**
+     * The actual cancellation state
+     */
+    private AtomicBoolean cancelled = new AtomicBoolean(false);
```
[the remark this answers is not in the document]
Comment (malaysf) [reply core/src/main/java/com/backblaze/b2/client/B2CancellationToken.java:20]: Honestly, I hadn't thought about it, spell checker didn't complain...

That said, Java has a `CancellationException` with two Ls. https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CancellationException.html

So I think it is best to use this spelling.

File: core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java [hunk 2]
```diff
@@ -55,6 +57,8 @@
      */
     private final List<Long> startingBytePositions;
 
+    final B2CancellationToken cancellationToken = new B2CancellationToken();
```
[the remark this answers is not in the document]
Comment (malaysf) [reply core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java:60]: oops, yes should be private thanks.

File: core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java [hunk 3]
```diff
@@ -179,17 +183,129 @@ B2FileVersion storeFile(B2UploadListener uploadListenerOrNull) throws B2Exceptio
...
         for (final B2PartStorer partStorer : partStorers) {
-            partFutures.add(executor.submit(() -> partStorer.storePart(this, uploadListener)));
+            partFutures.add(executor.submit(() -> partStorer.storePart(this, uploadListener, cancellationToken)));
         }
 
         return finishLargeFileFromB2PartFutures(fileVersion, partFutures);
     }
 
+    /**
+     * Stores the file asynchronously and returns a CompletionFuture to manage the task.
```
[the remark this answers is not in the document]
Comment (malaysf) [reply core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java:196]: yes thanks

File: core/src/main/java/com/backblaze/b2/client/B2CancellationToken.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            throw new B2LocalException("cancelled", "Request cancelled by caller");
+        }
+    }
+
+    /**
+     * Sets the cancelled state, will throw if already cancelled
+     * @throws B2Exception if previously cancelled
+     */
+    void cancel() throws B2Exception {
+        throwIfCancelled();
```
[the remark this answers is not in the document]
Comment (malaysf) [reply core/src/main/java/com/backblaze/b2/client/B2CancellationToken.java:37]: Good point. Future's definition of cancel also allows multiple invocations so I'll change this.

**Review by certainmagic — APPROVED** [review]
thanks.  these seem like good improvements.  :)

i'm wondering if we can/should have an example use somewhere -- as a javadoc comment somewhere? in B2Sample? in our README?  not sure where.

**Review by generictjohnson — APPROVED** [review]
Cool stuff! Just a few suggestions on the tests

File: core/src/test/java/com/backblaze/b2/client/B2CancellableContentSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    final B2CancellableContentSource cancellableContentSource = new B2CancellableContentSource(mockContentSource, cancellationToken);
+
+    @Before
+    public void setup() throws B2Exception, IOException {
+        when(mockContentSource.createInputStream()).thenReturn(mockInputStream);
+    }
+    @Test
+    public void testPassthroughMethods() throws B2Exception, IOException {
+        cancellableContentSource.getContentLength();
```
[the remark this answers is not in the document]
Comment (malaysf) [reply core/src/test/java/com/backblaze/b2/client/B2CancellableContentSourceTest.java:34]: great idea thanks!

File: core/src/test/java/com/backblaze/b2/client/B2CancellableContentSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        inputStream.reset();
+        verify(mockInputStream).reset();
+
+        inputStream.markSupported();
+        verify(mockInputStream).markSupported();
+    }
+
+    @Test(expected = IOException.class)
+    public void testInputStreamReadCancelled() throws B2Exception, IOException {
+        final InputStream inputStream = cancellableContentSource.createInputStream();
```
[the remark this answers is not in the document]
Comment (malaysf) [reply core/src/test/java/com/backblaze/b2/client/B2CancellableContentSourceTest.java:76]: you're not being annoying -- its an annoying change but the right one. Thanks for pointing this out.

**Pull request merged**